lib/browser/http.ts

```diff
@@ -193,3 +194,109 @@ export class HttpClientStream extends BufferedEventEmitter {
         this.emit('error', new Error(`msg=${error.message}, XHR=${error.request}, info=${info}`));
     }
 }
+
+interface PendingRequest {
+    resolve: (connection: HttpClientConnection) => void;
+    reject: (error: CrtError) => void;
+}
+
+/** Creates, manages, and vends connections to a given host/port endpoint */
+export class HttpClientConnectionManager {
+    private pending_connections = new Set<HttpClientConnection>();
+    private live_connections = new Set<HttpClientConnection>();
+    private free_connections: HttpClientConnection[] = [];
+    private pending_requests: PendingRequest[] = [];
+
+    constructor(
+        readonly host: string,
+        readonly port: number,
+        readonly max_connections: number
+    ) {
+        
+    }
+
+    private remove(connection: HttpClientConnection) {
+        this.pending_connections.delete(connection);
+        this.live_connections.delete(connection);
+        const free_idx = this.free_connections.indexOf(connection);
+        if (free_idx != -1) {
+            this.free_connections.splice(free_idx, 1);
+        }
+    }
+
+    private resolve(connection: HttpClientConnection) {
+        const request = this.pending_requests.shift();
+        if (request) {
+            request.resolve(connection);
+        }
```

Comment:
Feels like there should be an else that pushes the connection back on the free list

Reply:
I still think this needs an else.  Consider the following scenario

(1) Request a connection, create and establish a connection, start to use it
(2) Request a connection, create pending connection
(3) Finish with the connection in (1), return it, call pump and give it to the pending request from (2)
(4) The pending connection from (2) is established, it calls "this.resolve(connection)" but there's no pending requests and it promptly disappears.

Based on what you said about close() it seems like this would be "safe" but at the same time after going through all the effort of establishing a connection, we should keep it around rather than tossing it immediately.